### ISSUE TITLE
Auditing policies for 28 days results in error "Could not find a number of steps suitable for reporting period." 

### ISSUE THREAD

**Issue by sethhill** [seanhamlin/site-audit#246]
When running reports or individual policies with a duration of 28 days, error is returned "Could not find a number of steps suitable for reporting period.          
Query failed due to an unknown error. Please try again."

Sample command run:
`./vendor/bin/drutiny policy:audit AcquiaCS:SumoLogic:DrupalResponseCodes @drupalvm.dev --reporting-period-start 2020-04-01 --reporting-period-end 2020-04-29`

<img width="799" alt="Screen Shot 2020-05-18 at 07 34 59" src="https://user-images.githubusercontent.com/3334926/82225581-3ce2a400-98da-11ea-8f51-ab2e69bfb734.png">

**Comment by fiasco** [comment]
When running sumo queries, the query is sent to the API, then you have to poll the API to see if the result set is ready yet. Drutiny doesn't wait all day. It polls the API 200x then gives up and moves on to the next policy in the profile. This is probably what is happening here with a 28 day range. Not all customers experience this, it depends on the volume of data to process.

Drutiny could wait longer if it could do things in parallel so that other policies didn't have to wait for Sumologic to finish before they could start.

If you don't think this is the reason this policy failed. Please reopen with more information.

**Comment by sethhill** [comment]
Do I need to file a new issue, or can you reopen from here if you agree with my logic? This specific occurrence is unrelated to the Sumo API polling. I would expect that even with a fix here, users would still likely encounter the Sumo timeout with large data volume.

This issue happens because of the distribution of possible intervals in Sandbox/ReportingPeriodTrait, and the calculation of steps for timeslices/graphing. At 28 days, a 12-hour interval results in 56 steps; a 6-hour interval results in 112. This fails the logic in getReportingPeriodSteps(), which seeks to identify an interval for which steps ≥ 60 and ≤ 100. 

Report periods of 26–29 days should all fail similarly. Adding an 8-hour interval (28800 secs) would solve for this case.

But, anywhere that the ratio of one interval to the next is 1:2, there is a possibility for a reporting period that falls outside the desired number of steps. (For example, a report run with a duration of 30,600 seconds should similarly fail, as that would result in 51 x 600-sec steps, and 102 x 300-sec steps.)

I've encountered the 28-day bug myself, so would like to address that. Does it seem desirable/necessary to insert additional intervals to accommodate all possibilities?

**Comment by fiasco** [comment]
Ah yup this makes sense. @bschilt helped me understand this issue better too. So this isn't resolved with your last PR?

**Comment by fiasco** [comment]
I'm open to different math that will improve this, perhaps even some unit tests to cover which periods we know we support. 

The key is for the intervals to have some meaning to them so they're not things like "1 hour, 12 minutes and 53 seconds" (which you would get if you divided the time period by a fixed frequency) but are more interprible when viewed as graphs. Periods like 1min, 2min, 5min, 10min, 20mins 30mins, 1hr, 3hrs, 6hrs, 12hrs, 1 day make more sense when viewing the grainularity from a graph (they all round to an hour). But values like 45mins, 90mins, 9hrs, 15hrs are less desirable because they're harder to graphically read.

**Comment by sethhill** [comment]
I understand and agree with that logic. It also makes more sense if the smaller intervals line up with full days, etc., so the graphs have some repeatability of periods. I would advocate for changing the 5-hour interval to 4 hours for that reason.

Is there a specific logic for 60 steps minimum and 100 steps maximum, or was that a general calculation to strike a balance? I'll give the math some thought...

**Comment by fiasco** [comment]
> Is there a specific logic for 60 steps minimum and 100 steps maximum

Yes! this is an approximate range where graphs strike the best balance between high fidelity and browser performance. For example, a graph with 3 series of 100 steps each will have 300 plot points to graph which is doable for a browser. But a graph with 5 series and 200 steps each has 1000 plot points which is going to take 3x longer to render.

Since a report can contain many graphs, the performance required from the browser gets up their pretty quickly. So keeping the steps low allows policies to think less about these things and "just work".
